frontend/packages/console-shared/src/hooks/useUserSettings.ts

```diff
@@ -0,0 +1,106 @@
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore
+import { useSelector } from 'react-redux';
+import { RootState } from '@console/internal/redux';
+import { ConfigMapModel, ProjectRequestModel, ProjectModel } from '@console/internal/models';
+import { K8sResourceKind, k8sGet, k8sCreate, k8sPatch } from '@console/internal/module/k8s';
+import { useK8sWatchResource } from '@console/internal/components/utils/k8s-watch-hook';
+
+// can't create project with name prefix with 'openshift-*', once we have proxy need to update
+const USER_SETTING_CONFIGMAP_NAMESPACE = 'console-user-settings';
+
+const getConfigResourceData = (userUid: string, key: string, defaultValue: string) => {
+  return [
+    {
+      kind: ConfigMapModel.kind,
+      namespace: USER_SETTING_CONFIGMAP_NAMESPACE,
+      isList: false,
+      name: `user-settings-${userUid}`,
+    },
+    {
+      apiVersion: ConfigMapModel.apiVersion,
+      kind: ConfigMapModel.kind,
+      metadata: {
+        name: `user-settings-${userUid}`,
+        namespace: USER_SETTING_CONFIGMAP_NAMESPACE,
+      },
+      data: {
+        [key]: defaultValue,
+      },
+    },
+  ];
+};
+
+// This won't be needed once we have proxy api
+const getProject = async () => {
+  try {
+    await k8sGet(ProjectModel, USER_SETTING_CONFIGMAP_NAMESPACE);
+  } catch {
+    await k8sCreate(ProjectRequestModel, {
+      metadata: {
+        name: USER_SETTING_CONFIGMAP_NAMESPACE,
+      },
+    });
+  }
+};
+
+const createConfigMap = async (configMapData: K8sResourceKind) => {
+  try {
+    await k8sCreate(ConfigMapModel, configMapData);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(err);
+  }
+};
+
+const patchConfigMap = async (patch, cfData) => {
+  try {
+    await k8sPatch(ConfigMapModel, cfData, patch);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(err);
+  }
+};
+
+export const useUserSettings = (key: string, defaultValue: string = '') => {
+  const userUid = useSelector(
+    (state: RootState) => state.UI.get('user')?.metadata?.uid ?? 'kubeadmin',
+  );
+  const [configMapResource, configMapData] = getConfigResourceData(userUid, key, defaultValue);
+  const [cfData, cfLoaded, cfLoadError] = useK8sWatchResource<K8sResourceKind>(configMapResource);
+  const [settings, setSettings] = React.useState('');
+
+  React.useEffect(() => {
+    if (cfLoadError || (!cfData && cfLoaded)) {
+      (async () => {
+        await getProject();
+        await createConfigMap(configMapData);
+      })();
+    } else if (cfData && cfLoaded && settings !== cfData.data?.[key]) {
+      setSettings(cfData.data?.[key] ?? defaultValue);
+    }
+    // This effect should only be run once per key.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [key, cfLoaded]);
+
+  React.useEffect(() => {
+    if (cfData && cfLoaded && settings) {
+      const patch = [
+        {
+          op: cfData.data?.[key] ? 'replace' : 'add',
```

Comment:
```suggestion
          op: 'replace',
```

I though a little bit about parallel create user-settings and parallel patch calls. The `replace` operation works here for adding or updating while adding works just when the data key is not available.